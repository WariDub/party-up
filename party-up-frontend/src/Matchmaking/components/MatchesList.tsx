import { Box, Button, Card, CardContent, GridList, GridListTile } from '@material-ui/core';
import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import Match from '../models/match.model';
import * as axios from 'axios';
import * as jwtDecode from 'jwt-decode';
import User from '../../Profile/models/user.model';
import { BACKEND_URL } from '../../globals';
import { JwtPayload } from '../../Auth/interfaces/jwt-payload.interface';
import { UsersService } from '../../Profile/users.service';

export interface MatchesListProps extends RouteComponentProps {
  matches: Match[];
}

export interface MatchesListState {
  user: User | null;
}

const MatchesList = class extends React.PureComponent<MatchesListProps, MatchesListState> {
  usersService = new UsersService();

  constructor(props: MatchesListProps) {
    super(props);

    this.state = {
      user: null,
    };
  }

  async componentDidMount() {
    await this.fetchAndSetUser();
  }

  render(): JSX.Element {
    const { matches } = this.props;
    const { user } = this.state;

    return (
      <Box p={2}>
        <GridList spacing={16} cellHeight="auto" cols={0}>
          {user
            ? matches.map((match) => (
                <GridListTile key={match.entry.user._id} onClick={this.handleMatchOnClick(match)}>
                  {this.renderMatch(match)}
                </GridListTile>
              ))
            : []}
        </GridList>
      </Box>
    );
  }

  renderMatch(match: Match): JSX.Element {
    const { user } = this.state;

    return (
      <Card>
        <CardContent>
          <h3>{match.entry.user.displayName}</h3>
          <h5>{`In-game: ${match.entry.identifier}`}</h5>
          <h5>{`${match.percentage}% match`}</h5>
          <Button
            variant="contained"
            color="primary"
            onClick={this.handleButtonOnClickToggleFriend(match)}
          >
            {user?.friends.includes(match.entry.user._id) ? 'Unfriend' : 'Add Friend'}
          </Button>
        </CardContent>
      </Card>
    );
  }

  handleMatchOnClick = (param: Match) => (event: any): void => {
    console.log('match on click');
  };

  handleButtonOnClickToggleFriend = (param: Match) => (event: any): void => {
    const { user } = this.state;
    if (!user) {
      return;
    }

    const addFriend = !user.friends.includes(param.entry.user._id);

    if (addFriend) {
      user.friends.push(param.entry.user._id);
      event.target.innerHTML = 'Unfriend';
      this.usersService.addFriend(user.username, param.entry.user._id);
    } else {
      user.friends = user.friends.filter((friend) => friend != param.entry.user._id);
      event.target.innerHTML = 'Add Friend';
      this.usersService.deleteFriend(user.username, param.entry.user._id);
    }
  };

  fetchAndSetUser = async (): Promise<void> => {
    const accessToken = localStorage.getItem('token');
    if (!accessToken) {
      return console.error('missing access token');
    }

    const payload: JwtPayload = jwtDecode.default(accessToken);

    try {
      const user = await this.usersService.getUser(payload.uid);
      if (user) {
        this.setState({ user });
      }
    } catch (e) {
      console.error(e);
    }
  };
};

export default MatchesList;
