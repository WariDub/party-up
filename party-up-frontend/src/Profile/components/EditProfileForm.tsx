import {
  Box,
  Button,
  Card,
  CardActions,
  CardContent,
  FormControl,
  Grid,
  InputLabel,
  MenuItem,
  Select,
  TextField,
  Typography,
} from '@material-ui/core';
import React from 'react';
import Gender from '../enums/gender.enum';
import Genre from '../enums/genre.enum';
import Role from '../enums/role.enum';

export interface EditProfileFormProps {}

export interface EditProfileFormState {
  age: number;
  gender: Gender;
  favoriteGenre: Genre;
  favoriteRole: Role;
}

const EditProfileForm = class extends React.Component<EditProfileFormProps, EditProfileFormState> {
  constructor(props: EditProfileFormProps) {
    super(props);

    this.state = {
      age: 0,
      gender: Gender.OTHER,
      favoriteGenre: Genre.ACTION,
      favoriteRole: Role.DPS,
    };
  }

  render(): JSX.Element {
    return (
      <Box m={1}>
        <Grid container justify="center">
          <Grid item xs={8}>
            {this.renderCard()}
          </Grid>
        </Grid>
      </Box>
    );
  }

  renderCard(): JSX.Element {
    return (
      <Card elevation={3}>
        <CardContent>
          <Grid container justify="center" spacing={1}>
            <Grid item xs={10}>
              <Typography variant="h6">Profile</Typography>
            </Grid>
            {this.renderQuestionAge()}
            {this.renderQuestionGender()}
            {this.renderQuestionFavoriteGenre()}
            {this.renderQuestionFavoriteRole()}
          </Grid>
        </CardContent>
        <CardActions>
          <Grid container justify="center" spacing={1}>
            <Grid item>
              <Button size="medium" onClick={this.handleButtonOnClickSaveChanges}>
                Save Changes
              </Button>
            </Grid>
          </Grid>
        </CardActions>
      </Card>
    );
  }

  renderQuestionAge(): JSX.Element {
    return (
      <Grid item xs={10}>
        <FormControl fullWidth variant="filled">
          <TextField variant="filled" label="Age" onChange={this.handleChangeAge} />
        </FormControl>
      </Grid>
    );
  }

  handleChangeAge = (event: React.ChangeEvent<{ value: unknown }>): void => {
    const value = event.target.value as number;
    if (value > 0) {
      this.setState({ age: value });
    }
  };

  // eslint-disable-next-line class-methods-use-this
  renderFieldWithChoices(
    label: string,
    value: string | number,
    choices: { [key: string]: string | number },
    onChange: (event: React.ChangeEvent<{ value: unknown }>) => void
  ): JSX.Element {
    return (
      <Grid item xs={10}>
        <FormControl fullWidth variant="filled">
          <InputLabel>{label}</InputLabel>
          <Select value={value} onChange={onChange}>
            {Object.values(choices).map((choice) => (
              <MenuItem key={choice} value={choice}>
                {choice}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      </Grid>
    );
  }

  renderQuestionGender(): JSX.Element {
    const { gender } = this.state;
    return this.renderFieldWithChoices('Gender', gender, Gender, this.handleChangeGender);
  }

  renderQuestionFavoriteGenre(): JSX.Element {
    const { favoriteGenre } = this.state;
    return this.renderFieldWithChoices(
      'Favorite Genre',
      favoriteGenre,
      Genre,
      this.handleChangeFavoriteGenre
    );
  }

  renderQuestionFavoriteRole(): JSX.Element {
    const { favoriteRole } = this.state;
    return this.renderFieldWithChoices(
      'Favorite Role',
      favoriteRole,
      Role,
      this.handleChangeFavoriteRole
    );
  }

  handleChangeGender = (event: React.ChangeEvent<{ value: unknown }>): void => {
    const value = event.target.value as Gender;
    this.setState({ gender: value });
  };

  handleChangeFavoriteGenre = (event: React.ChangeEvent<{ value: unknown }>): void => {
    const value = event.target.value as Genre;
    this.setState({ favoriteGenre: value });
  };

  handleChangeFavoriteRole = (event: React.ChangeEvent<{ value: unknown }>): void => {
    const value = event.target.value as Role;
    this.setState({ favoriteRole: value });
  };

  handleButtonOnClickSaveChanges = async (): Promise<void> => {
    const { age, gender, favoriteGenre, favoriteRole } = this.state;
    console.log(age, gender, favoriteGenre, favoriteRole);
    // TODO: connect to backend to save user profile
  };
};

export default EditProfileForm;
