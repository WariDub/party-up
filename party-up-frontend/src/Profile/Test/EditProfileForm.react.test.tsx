import { shallow } from 'enzyme';
import renderer from 'react-test-renderer'
import  EditProfileForm  from '../components/EditProfileForm';
import User from '../models/user.model';
import Gender from '../enums/gender.enum';
import Genre from '../enums/genre.enum';
import Role from '../enums/role.enum';
import Match from '../../Matchmaking/models/match.model';
import Experience from '../../Matchmaking/enums/experience.enum';
import MatchMakingEntryModel from '../../Matchmaking/models/matchmaking-entry.model';
import MatchesList from '../../Matchmaking/components/MatchesList';
import { createBrowserHistory } from 'history';
import { matchPath } from 'react-router';


describe('EditProfileForm', () => {
    test('It should mount', () => {
        let history = createBrowserHistory();

        let testUser: User = {
            _id: '1',
            username: 'TestUsername',
            displayName: 'TestDisplayName',
            email: 'TestEmail@TestEmail.com',
            avatarUrl: 'TestAvatarURL.com',
            friends: ['Friend1', 'Friend2'],
            age: 23,
            gender: Gender.MALE,
            favoriteGenre: Genre.ACTION,
            favoriteRole: Role.DPS,
            createdAt: 5/23/2021
        }

        history.push('/');

        let location = history.location;
        
        let matchity = matchPath('/ProfilePage', {
            path: '/ProfilePage',
            exact: true,
            strict: false,
        });

        let component = shallow(<EditProfileForm user={testUser} match={ matchity } history= { history } location = { location } />)

        expect(component.length).toBe(1);
    });

    test('It should render the EditProfileForm, and display all of the appropriate fields', () => {
        let history = createBrowserHistory();

        let testUser: User = {
            _id: '1',
            username: 'TestUsername',
            displayName: 'TestDisplayName',
            email: 'TestEmail@TestEmail.com',
            avatarUrl: 'TestAvatarURL.com',
            friends: ['Friend1', 'Friend2'],
            age: 23,
            gender: Gender.MALE,
            favoriteGenre: Genre.ACTION,
            favoriteRole: Role.DPS,
            createdAt: 5/23/2021
        }

        history.push('/');

        let location = history.location;
        
        let matchity = matchPath('/ProfilePage', {
            path: '/ProfilePage',
            exact: true,
            strict: false,
        });

        let component = renderer.create(<EditProfileForm user={testUser} match={ matchity } history= { history } location = { location } />).toJSON();


        expect(component).toMatchSnapshot();
    });


});