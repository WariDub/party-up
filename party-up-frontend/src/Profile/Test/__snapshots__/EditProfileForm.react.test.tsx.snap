// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfileForm It should render the EditProfileForm, and display all of the appropriate fields 1`] = `
<div
  className="MuiBox-root MuiBox-root-1"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded"
      >
        <div
          className="MuiCardContent-root"
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-justify-xs-center"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
            >
              <h6
                className="MuiTypography-root MuiTypography-h6"
              >
                Profile
              </h6>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
            >
              <div
                className="MuiFormControl-root MuiFormControl-fullWidth"
              >
                <div
                  className="MuiFormControl-root MuiTextField-root"
                >
                  <label
                    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
                    data-shrink={true}
                  >
                    Age
                  </label>
                  <div
                    className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                    onClick={[Function]}
                  >
                    <input
                      aria-invalid={false}
                      autoFocus={false}
                      className="MuiInputBase-input MuiFilledInput-input"
                      disabled={false}
                      onAnimationStart={[Function]}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      required={false}
                      type="text"
                      value={23}
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
            >
              <div
                className="MuiFormControl-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
                  data-shrink={true}
                >
                  Gender
                </label>
                <div
                  className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <div
                    aria-haspopup="listbox"
                    className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    Male
                  </div>
                  <input
                    aria-hidden={true}
                    className="MuiSelect-nativeInput"
                    onAnimationStart={[Function]}
                    onChange={[Function]}
                    required={false}
                    tabIndex={-1}
                    value="Male"
                  />
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
            >
              <div
                className="MuiFormControl-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
                  data-shrink={true}
                >
                  Favorite Genre
                </label>
                <div
                  className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <div
                    aria-haspopup="listbox"
                    className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    Action
                  </div>
                  <input
                    aria-hidden={true}
                    className="MuiSelect-nativeInput"
                    onAnimationStart={[Function]}
                    onChange={[Function]}
                    required={false}
                    tabIndex={-1}
                    value="Action"
                  />
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
            >
              <div
                className="MuiFormControl-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
                  data-shrink={true}
                >
                  Favorite Role
                </label>
                <div
                  className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <div
                    aria-haspopup="listbox"
                    className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    DPS
                  </div>
                  <input
                    aria-hidden={true}
                    className="MuiSelect-nativeInput"
                    onAnimationStart={[Function]}
                    onChange={[Function]}
                    required={false}
                    tabIndex={-1}
                    value="DPS"
                  />
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="MuiCardActions-root MuiCardActions-spacing"
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-justify-xs-center"
          >
            <div
              className="MuiGrid-root MuiGrid-item"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-text"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Save Changes
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
